### PR TITLE
Server routine: Process dirty topics sub-routine

### PR DESCRIPTION
**Pull request by EduPonz**
Merge after #1410

### CODE REVIEW

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 1]
```diff
@@ -328,7 +328,7 @@ bool PDPServer2::server_update_routine()
     bool result = process_writers_acknowledgements();  // server + ddb(functor_with_ddb)
     // result |= process_data_queue();                    // all ddb
     result |= process_disposals();                     // server + ddb(get_disposals, clear_changes_to_disposes)
-    // result |= process_dirty_topics();                  // all ddb
+    result |= discovery_db_.process_dirty_topics();    // all ddb
```
Comment (jparisu) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:331]:
```suggestion
    result |= process_dirty_topics();    // all ddb
```
I would suggest to create a new method in PDPServer2. It is not necessary but keeps the logic uniform along the process methods

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp [hunk 2]
```diff
@@ -156,7 +156,123 @@ bool DiscoveryDataBase::process_data_queue()
...
+        }
+        else
+        {
+            // Mark are_dirty_topics as true since this topic is still dirty
+            are_dirty_topics = true;
+            // Proceed with next topic
+            ++topic_it;
+        }
+    }
+    return are_dirty_topics;
```
Comment (jparisu) [inline src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp:275]:
```suggestion
    return dirty_topics_.empty();
```
you can avoid using an extra boolean

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp [hunk 3]
```diff
@@ -156,7 +156,123 @@ bool DiscoveryDataBase::process_data_queue()
...
+                    // Set topic as not-clearable.
+                    is_clearable = false;
+                }
+
+                // Check in `participants_` whether the client with the writer has acknowledge the PDP of the client
+                // with the reader.
+                if (parts_writer_it != participants_.end() && parts_writer_it->second.is_matched(reader.guidPrefix))
+                {
+                    // Check the status of the reader in `writers_[writer]::relevant_participants_builtin_ack_status`.
+                    auto writers_it = writers_.find(writer);
```
Comment (jparisu) [inline src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp:232]: Erase this line. These declarations has been moved to the beginning of the function.

**Review by jparisu — CHANGES_REQUESTED** [review]
Small changes and uncrustify needed.

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp (hunk 1 above)
Comment (EduPonz) [reply src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:331]: :+1:

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp (hunk 2 above)
Comment (EduPonz) [reply src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp:275]: Good catch :robot:

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp (hunk 3 above)
Comment (EduPonz) [reply src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp:232]: :+1:

**Review by jparisu — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2698/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/91/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2495/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2708/)

**Pull request merged**